src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,668 @@
+        batch.insert(
+            PART_ID,
+            vec![
+                (2, MzOffset::from(2)),
+                (3, MzOffset::from(3)),
+                (4, MzOffset::from(4)),
+            ],
+        );
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(2, 1000), (3, 2000), (4, 2000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Advance forward beyond the first binding
+        upper.insert(PART_ID, MzOffset::from(3));
+        assert_eq!(Antichain::from_elem(2000), operator.advance(&upper).await);
+
+        // We haven't compacted so we should still be able to correctly reclock timestamps < 3
+        batch.insert(PART_ID, vec![(2, MzOffset::from(2))]);
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(2, 1000)]);
+        assert!(batch[&PART_ID].is_empty());
+    }
+
+    #[tokio::test(start_paused = true)]
+    #[should_panic(expected = "insufficient binding history")]
+    async fn test_compact() {
+        const PART_ID: PartitionId = PartitionId::None;
+        let mut operator = test_operator(ShardId::new()).await;
+
+        let mut batch = HashMap::new();
+
+        // Reclock offsets 1 and 2 to timestamp 1000
+        batch.insert(
+            PART_ID,
+            vec![(1, MzOffset::from(1)), (2, MzOffset::from(2))],
+        );
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(1, 1000), (2, 1000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Reclock offsets 3 and 4 to timestamp 2000
+        batch.insert(
+            PART_ID,
+            vec![(3, MzOffset::from(3)), (4, MzOffset::from(4))],
+        );
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(3, 2000), (4, 2000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Compact enough so that we can correctly timestamp only offsets > 2
+        operator.compact(Antichain::from_elem(1000)).await;
+
+        // Reclock offsets 3 and 4 again to see we haven't lost the ability
+        batch.insert(
+            PART_ID,
+            vec![(3, MzOffset::from(3)), (4, MzOffset::from(4))],
+        );
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(3, 2000), (4, 2000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Attempt to reclock offset 2 which should cause a panic
```

Comment:
if you improve the panic message in my suggestion above, we can make this test more reliable by making the expected panic message the full message with the erroneous offset

Reply:
Changed these to Results for better testing